torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -201,10 +200,62 @@ class ShapePropagator {
         continue;
       }
     }
-
     return tensor_types;
   }
 
+  c10::ScalarType unionScalarTypes(c10::ScalarType original, c10::ScalarType next) {
+    if (original == c10::ScalarType::Undefined) {
+      return next;
+    } else {
+      return c10::promoteTypes(original, next);
+    }
+  }
+
+  // Promotes result types for arithmetic operations on Tensor operands using
+  // new type promotion logic. See tensor_attributes.rst for details.
+  // This doesn't handle the case of arithmetic ops with Scalar arguments (when
+  // `Tensor.getUnsafeTensorImpl()->is_wrapped_nubmer()` would return true)
+  c10::optional<c10::ScalarType> getPromotedTypeForArithmeticOp(Node *node) {
+    c10::ScalarType dimmed = c10::ScalarType::Undefined;
+    c10::ScalarType zerodim = c10::ScalarType::Undefined;
+    // binary arithmetic ops, more than 2 args is alpha.
+    for (size_t i = 0 ; i < 2 ; i++ ) {
+      auto dtt = node->inputs()[i]->type()->expect<ProfiledTensorType>();
+      auto inputDtype = dtt->scalarType();
+      if (!dtt || !inputDtype) {
+        return c10::nullopt;
```

Comment:
it seems to me like this should never happen, but scalarType() returns an optional. Not sure if an assert would be better?